WIP Gemcalibration package

cgalloni opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Work in progress : branch for calibration suite development

## Description
<!--- Describe your changes in detail -->
Creation of calibration suite 

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue(s) here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

### Screenshots (if appropriate):

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--- Template thanks to https://www.talater.com/open-source-templates/#/page/99 -->

cgalloni committed:
inital commit  gemcalibration

cgalloni committed:
cal routines main interface, with AJAX requests.

cgalloni committed:
inital commit  gemcalibration

cgalloni committed:
resolving remaining bugs

cgalloni committed:
pushing change to xml

cgalloni committed:
revert change to xml

cgalloni committed:
changing host in xml

cgalloni committed:
inital web interface for scurve calib

cgalloni commented:
The interface at the momet looks like this. Suggestions are very welcome. 

![screen shot 2019-02-27 at 5 07 37 pm](https://user-images.githubusercontent.com/6152470/53504887-19ec8a00-3ab3-11e9-80de-5832f72434e3.png)

bdorney commented:
> The interface at the momet looks like this. Suggestions are very welcome.
> 
> ![screen shot 2019-02-27 at 5 07 37 pm](https://user-images.githubusercontent.com/6152470/53504887-19ec8a00-3ab3-11e9-80de-5832f72434e3.png)

So here the AMC13 will be generating the BGO commands (L1A and calpulse).  In the current tool we use the TTC Generator block on the CTP7 (not the AMC13) and we can set the period of calpulse commands and the delay between the the L1A and the calpulse.

However how can you do this here?  I don't see a field for calpulse period or pulse delay between L1A and calpulse.

Also instead of "Card Name" I would go by (shelf,slot) as it is in the cmsgemos connections file.

Also scurve's aren't "calibration" per say so I would change title to "Scurve Interface" or something similar.

bdorney commented:
Also what is the use case here for "vfat mask"?  This is to be applied to all optohybrids in OH mask?

I think if you are going to let the user give an OH mask then you should not let them give a vfat mask and instead this should be determined automatically (there exists an rpc module to do this...whether it is yet used by the xdaq interface @jsturdy and @mexanick could comment).  The reason being that each OH could need a different vfat mask.

bdorney commented:
Another comment, VFAT channel is defined as [0,127] and the channel 128 doesn't exist.  See for example [this document](https://edms.cern.ch/ui/#!master/navigator/document?P:1115513923:100178398:subDocs) which describes the channel mapping for HV3b_V1, V2, and V4 hybrids.

jsturdy commented:
Question: How does the list of AMCs get populated before you get to this point?

Suggestions:
* A button that can "apply to all" settings can be useful, in the case that most settings will be common, and individual ones can be changed afterwards.
* Some of the settings you list are very unlikely to be different, e.g., number of samples, and should be implemented in a more "scan global" part

mexanick commented on gemcalibration/src/common/CalibrationWeb.cc:
```diff
@@ -0,0 +1,500 @@
+ *out << "  <td> "<< std::endl;
+ *out << "<div class=\"checkbox\">"<< std::endl;
+ *out << "<label>"<< std::endl;
+ *out << "<input type=\"checkbox\" class=\"check\"> eagleB" << std::endl;
```
This line and all code below seems to be a carbon copy of the `eagleA` table definition above. Please refactor this using the `for` loop, looping over the CTP7's boards in question.
Also, the input fields should receive unique identifiers so that they can be distinguished in a callback function

mexanick commented on gemcalibration/html/scripts/shelfMonitor.js:
```diff
@@ -0,0 +1,92 @@
+var actionURL = "";
```
Please remove this file from the project. You're not using it and will have to write your own script later.

mexanick reviewed (dismissed):
Considering the interface: as we discussed, a possibility to work with multiple crates has to be foreseen. While `eagle`s are fine for the moment, I think we should use geographic ID's in the future. @jsturdy could you please provide your thoughts on this. 
Another thing to discuss: eventually we should move the `css` to a common place, as more than one sub package is using it. There's no point to keep them in all the sub packages.

jsturdy commented:
> Another thing to discuss: eventually we should move the `css` to a common place, as more than one sub package is using it. There's no point to keep them in all the sub packages.

If it's used in multiple packages, feel free to migrate it to `gembase`, which is a dependency of *all* web applications

bdorney commented:
Also one final thought would be, all these input parameters, e.g.:

- chMin,
- chMax,
- nevt,
- lat,
- pulse stretch, etc...

I would rather have the dialog boxes at the top and these applied to all selected AMC's rather than having to fill out all these parameters for all selected AMC's.

The only *per* AMC dialog box that I would have would be `ohMask`.

jsturdy commented on gemcalibration/Makefile:
```diff
@@ -0,0 +1,57 @@
+Package=gemcalibration
+LongPackage=gemcalibration
+
+ShortPackage=gemcal
```
`ShortPackage` (and namespaces) should be `calib`, i.e., 
* `gem::calib::blah`
* `include "gem/calib/file.h"`
* dir structure `include/gem/calib`

jsturdy commented on gemutils/include/gem/utils/db/GEMDatabaseUtils.h:
```diff
@@ -1,7 +1,7 @@
 #define GEM_UTILS_DB_GEMDATABASEUTILS_H
 
-#include "Python.h"
+#include "python2.7/Python.h"
```
Revert, i.e., `rebase` and exclude the commit that added this.
If you're having a problem building, please use the mattermost `CMS GEM DAQ/Online Software` channel, or open an issue

jsturdy commented:
One other suggestion that I had discussed with @mexanick when setting up this task, was to start with the scans that don't rely on tracking data (e.g., trigger rate scan with s-bits, DAC scans, etc.,), as they will require less additional setup.
The UI that you have started designing here can be reused, and in fact, it might even be worthwhile to template it a little where the scan type can be selected and then the fields may change slightly, as per the necessities of the individual scans.

lmoureaux committed:
Fix gemonlinedb build (#255)

* Fix `#include`s
  * `AMCConfiguration.h`
  * `OHV3Configuration.h`
  * `VFAT3ChipConfiguration.h`
* Add missing file `AMC13Configuration.xsd`
  * Reported by @lpetre-ulb, discovered thanks to tests.

lmoureaux committed:
Fix gemonlinedb build (#255)

* Fix `#include`s
  * `AMCConfiguration.h`
  * `OHV3Configuration.h`
  * `VFAT3ChipConfiguration.h`
* Add missing file `AMC13Configuration.xsd`
  * Reported by @lpetre-ulb, discovered thanks to tests.

Update XHAL path

Change expected `xhal` path from `XHALROOT` to `XHAL_ROOT`,
to be consistent with other packages and `etc/profile.d/gemdaqenv.sh`

gemcalibration module structure

inital commit  gemcalibration

Squashing intial commits for gemcalibration

jsturdy committed:
Update XHAL path

Change expected `xhal` path from `XHALROOT` to `XHAL_ROOT`,
to be consistent with other packages and `etc/profile.d/gemdaqenv.sh`

mexanick committed:
Merge pull request #256 from jsturdy/hotfix/xhal-path-setup

Update XHAL path

cgalloni committed:
attempt to use  dropdown

Commit:
Merge remote-tracking branch 'camilla/gemcalib-cami' into develop

Commit:
Reimplementing jsonUpdate. Need to check why base implementation didn't work

cgalloni committed:
Merge pull request #1 from mexanick/feature-helpCamilla

Feature help camilla

cgalloni committed:
web interface for scurve

cgalloni committed:
latency scan tentative interface

cgalloni committed:
in development

cgalloni committed:
attempt to getvalue from post

Commit:
Reworking the cal routines main interface, implementing AJAX requests. WIP: needs addition of other than s-curve interfaces and passing all the required parameters (for the moment it takes only few, as an example

Commit:
Minor

cgalloni committed:
Merge pull request #2 from mexanick/feature-helpCamilla

Reworking the cal routines main interface, implementing AJAX requests…

cgalloni committed:
adding other calib parameters

cgalloni committed:
adding latency scan interface

cgalloni committed:
correcting typo in printout

cgalloni committed:
fixing variable definistion

mexanick committed:
Merge pull request #3 from cgalloni/gemcalib-cami

Gemcalib cami

cgalloni committed:
adding thrDAC scan interface

cgalloni committed:
changing tabs to space

Commit:
Slightly updating the interface and implemented the scan parameters passing. HW selection is still TODO

cgalloni committed:
adding other interface

Commit:
Merge new interfaces, small refactoring

cgalloni committed:
Merge pull request #3 from mexanick/feature-gemcalibration

Feature gemcalibration

Commit:
Interface refactoring based on calibration type maps

cgalloni committed:
Merge pull request #4 from mexanick/feature-gemcalibration

Interface refactoring based on calibration type maps

mexanick committed:
Adding propagation of slots and links selection

mexanick committed:
Moved bootstrap css and js to gembase

cgalloni committed:
Merge pull request #5 from mexanick/feature-gemcalibration

Adding propagation of slots and links selection

cgalloni committed:
partial integration of interface comments

cgalloni committed:
changing cal_type names, removing temperature

cgalloni committed:
removing SBITREADOUT,SBITMAPANDRATE

cgalloni committed:
solving merging conflicts

lmoureaux commented:
Commit 71fafc08673f4d459fb6cb3fcea5108a7b1f921d is weird: I am quoted as the author, it has the same summary as a commit already in `develop` and has changes spread across `gemonlinedb`, many `Makefile`s and `gemcalibration`.

Please at least edit the commit to change the author and commit message (perform an interactive rebase and use `git commit --amend --author "Name <email>"`).

mexanick commented:
@lmoureaux , @cgalloni is trying to perform her first rebase/squash. It could be done not 100% correctly, thus having some extra commits in a pipeline. We will figure this out.

lmoureaux commented:
np, sorry that I was a bit aggressive with this commit... it bugged me a lot to see my name on top of changes that weren't mine.

jsturdy commented:
The issue is probably due to the fact that this was `rebase`d onto `develop` when it should not have been.
(*or* because during the course of development, an incorrect merge was made, pulling in features that were not a part of `feature/gemcalib`, which is the base and target for PRs working on the calibration suite)
Tomorrow I'll go through the tree and suggest a corrective action

cgalloni commented:
@lmoureaux and @jsturdy , yes sorry, my fault. I made the `rebase` to the wrong branch, `develop`, which was the starting point of the implementation, while I should have done it to  `feature/gemcalib`.

jsturdy commented:
Here is the branch history:
```bash
*   6d14135      (HEAD -> camilla-gemcalib-rebased, cgalloni/gemcalib-cami) solving merging conflicts
|\  
| * fac3430       removing SBITREADOUT,SBITMAPANDRATE
| * 1b50138      changing cal_type names, removing temperature
| * 47977dd      partial integration of interface comments
| *   55d121d    Merge pull request #5 from mexanick/feature-gemcalibration
| |\  
| | * 49ad301    Moved bootstrap css and js to gembase
| | * bcb7960    Adding propagation of slots and links selection
| * |   1afb62a  Merge pull request #4 from mexanick/feature-gemcalibration
| |\ \  
| | |/  
| | * a4d68ec    Interface refactoring based on calibration type maps
| * |   3d5d67b  Merge pull request #3 from mexanick/feature-gemcalibration
| |\ \  
| | |/  
| | *   bdf4bd1  Merge new interfaces, small refactoring
| | |\  
| | |/  
| |/|   
| * | 7fc6c81    adding other interface
| * | 749db4c    changing tabs to space
| * | e892bcd    adding thrDAC scan interface
| * | f47179f     fixing variable definistion
| | * 7c3413a    Slightly updating the interface and implemented the scan parameters passing. HW selection is still TODO
| | *   88cbaa6  (mexanick/feature-helpCamilla) Merge pull request #3 from cgalloni/gemcalib-cami
| | |\  
| | |/  
| |/|   
| * | 62b92e1    correcting typo in printout
| * | 6c9ae11    adding latency scan interface
| * | 21910fe     adding other calib parameters
| * |   1725114  (gemcalib-cami) Merge pull request #2 from mexanick/feature-helpCamilla
| |\ \  
| | |/  
| | * ea1ea86    Minor
| | * 9ceccf4    Reworking the cal routines main interface, implementing AJAX requests. WIP: needs addition of other than s-curve interfaces and passing all the required parameters (for the moment it takes only few, as an example
| |/  
| * 59ed8cc      attempt to getvalue from post
| * 71cff34      in development
| * c99c068      latency scan tentative interface
| * a3e58a5      web interface for scurve
| *   964c4ce    Merge pull request #1 from mexanick/feature-helpCamilla
| |\  
| | * 8c3fa44    Reimplementing jsonUpdate. Need to check why base implementation didn't work
| | *   6e0cde7  Merge remote-tracking branch 'camilla/gemcalib-cami' into develop
| | |\  
| | |/  
| |/|   
| * | f463fa8    attempt to use  dropdown
| * | 3d0d89b    inital web interface for scurve calib
| * | 4104345    changing host in xml
| * | 5ea040c    revert change to xml
| * | 1586888    pushing change to xml
| * | 4805f5d     resolving remaining bugs
| * | 579d0cf    inital commit  gemcalibration
| * | 6f90292    inital commit  gemcalibration
| * | aa78f4a    (gitlab/feature/gemcalib, gemdaq/feature/gemcalib, feature/gemcalib) gemcalibration module structure
| | *   2c78ae1  Merge pull request #256 from jsturdy/hotfix/xhal-path-setup
| | |\  
| | | * bcb719e  Update XHAL path
| | |/  
| | * 19d9077    Fix gemonlinedb build (#255)
| |/  
* | 2db6cf2      cal routines main interface, with AJAX requests.
* | 71fafc0      Fix gemonlinedb build (#255)
|/  
*   1fdff44      (louis/develop) Merge pull request #245 from jsturdy/feature/add-xhal-interfaces-to-hwdevices
```

You can see that this branch was merged into a commit (`1fdff44`) that corresponds to an old pointer to the `develop` branch (this part may not be clear from this, but `louis/develop` is the `develop` branch in @lmoureaux's fork, which is not current with the `cms-gem-daq-project/develop` branch).
However, the development of these features was started (correctly) from commit `aa78f4a`, which corresponds to the `feature/gemcalib` branch.

Another problem is this commit:
```
* | 2db6cf2      cal routines main interface, with AJAX requests.
```
which was incorrectly made onto the (then) `develop` branch, rather than the `feature/gemcalib` branch, and in turn pulled in `71fafc0` and gave it a new hash (probably as part of a `rebase` `squash`)

From the full tree, 
```bash
| *   6d14135	 (HEAD -> camilla-gemcalib-rebased, cgalloni/gemcalib-cami) solving merging conflicts
| |\  
| * | 2db6cf2	 cal routines main interface, with AJAX requests.
| * | 71fafc0	 Fix gemonlinedb build (#255)
| | | | | *   cbbfe56	 (mexanick/feature-gemcalibration) Merge pull request #4 from cgalloni/gemcalib-cami
| | | | | |\  
| | | |_|_|/  
| | |/| | |   
| | * | | | fac3430	  removing SBITREADOUT,SBITMAPANDRATE
| | * | | | 1b50138	 changing cal_type names, removing temperature
| | * | | | 47977dd	 partial integration of interface comments
| | * | | |   55d121d	 Merge pull request #5 from mexanick/feature-gemcalibration
| | |\ \ \ \  
| | | | |_|/  
| | | |/| |   
| | | * | | 49ad301	 Moved bootstrap css and js to gembase
| | | * | | bcb7960	 Adding propagation of slots and links selection
| | * | | |   1afb62a	 Merge pull request #4 from mexanick/feature-gemcalibration
| | |\ \ \ \  
| | | |/ / /  
| | | * | | a4d68ec	 Interface refactoring based on calibration type maps
| | * | | |   3d5d67b	 Merge pull request #3 from mexanick/feature-gemcalibration
| | |\ \ \ \  
| | | |/ / /  
| | | * | |   bdf4bd1	 Merge new interfaces, small refactoring
| | | |\ \ \  
| | | |/ / /  
| | |/| | |   
| | * | | | 7fc6c81	 adding other interface
| | * | | | 749db4c	 changing tabs to space
| | * | | | e892bcd	 adding thrDAC scan interface
| | * | | | f47179f	  fixing variable definistion
| | | * | | 7c3413a	 Slightly updating the interface and implemented the scan parameters passing. HW selection is still TODO
| | | * | |   88cbaa6	 (mexanick/feature-helpCamilla) Merge pull request #3 from cgalloni/gemcalib-cami
| | | |\ \ \  
| | | |/ / /  
| | |/| | |   
| | * | | | 62b92e1	 correcting typo in printout
| | * | | | 6c9ae11	 adding latency scan interface
| | * | | | 21910fe	  adding other calib parameters
| | * | | |   1725114	 (gemcalib-cami) Merge pull request #2 from mexanick/feature-helpCamilla
| | |\ \ \ \  
| | | |/ / /  
| | | * | | ea1ea86	 Minor
| | | * | | 9ceccf4	 Reworking the cal routines main interface, implementing AJAX requests. WIP: needs addition of other than s-curve interfaces and passing all the required parameters (for the moment it takes only few, as an example
| | |/ / /  
| | * | | 59ed8cc	 attempt to getvalue from post
| | * | | 71cff34	 in development
| | * | | c99c068	 latency scan tentative interface
| | * | | a3e58a5	 web interface for scurve
| | * | |   964c4ce	 Merge pull request #1 from mexanick/feature-helpCamilla
| | |\ \ \  
| | | * | | 8c3fa44	 Reimplementing jsonUpdate. Need to check why base implementation didn't work
| | | * | |   6e0cde7	 Merge remote-tracking branch 'camilla/gemcalib-cami' into develop
| | | |\ \ \  
| | | |/ / /  
| | |/| | |   
| | * | | | f463fa8	 attempt to use  dropdown
| | * | | | 3d0d89b	 inital web interface for scurve calib
| | * | | | 4104345	 changing host in xml
| | * | | | 5ea040c	 revert change to xml
| | * | | | 1586888	 pushing change to xml
| | * | | | 4805f5d	  resolving remaining bugs
| | * | | | 579d0cf	 inital commit  gemcalibration
| | * | | | 6f90292	 inital commit  gemcalibration
| | * | | | aa78f4a	 (gitlab/feature/gemcalib, gemdaq/feature/gemcalib, feature/gemcalib) gemcalibration module structure
| |/ / / /  
* | | | |   f86cc48	 (gitlab/develop, gemdaq/develop, feature/boost-python-bindings-update, develop) Merge pull request #266 from lmoureaux/feature/onlinedb-manager
|\ \ \ \ \  
| |_|/ / /  
|/| | | |   
| * | | | dfa28db	 Add FIXME for Jared
| * | | | 861589e	 Fix style of Doxygen comments
| * | | | c7b119d	 Use /opt/cmsgemos/etc instead of /etc/opt/cmsgemos
| * | | | c3eef5f	 Drop unused onlinedb exceptions
| * | | | 2c81c32	 Make XercesGuard thread-safe
| * | | | 03618c0	 Update gemonlinedb.h
| * | | | 78d41f2	 Simplify and document exceptions in ConfigurationManager
| * | | | 97458ed	 Use Xerces exceptions wrapper in tests
| * | | | 1d8d54d	 Make it easier to catch Xerces exceptions
| * | | | b6abb1c	 Make configuration access thread-safe
| * | | | a9f4e80	 Add gem::onlinedb::ConfigurationManager
| * | | | 95bf3b1	 Make getFileInPath return absolute paths
|/ / / /  
* | | |   2c78ae1	 Merge pull request #256 from jsturdy/hotfix/xhal-path-setup
|\ \ \ \  
| * | | | bcb719e	 Update XHAL path
|/ / / /  
* | | | 19d9077	 Fix gemonlinedb build (#255)
|/ / /  
| | | * a1e57dd	 (gitlab/citest/gitlab-integration) ok, stop for now, need to get the external deps included in a sensible way...
| | | * 0666f81	 # This is a combination of 2 commits. # The first commit's message is:
| | | * 4a815ea	 broke the compile job, yaml shouldn't have created a valid job...
| | | * 24560d8	 add ability to test CI/CD based on branch name
| | | * b409d5c	 add EOS deploy script, updated gitlab CI yaml
| |_|/  
|/| |   
* | |   1fdff44	 (louis/develop) Merge pull request #245 from jsturdy/feature/add-xhal-interfaces-to-hwdevices
```
you can see that this comes from squashing commits into `19d9077`

jsturdy commented:
@cgalloni, after several attempts to clean this up, you are probably right, the simplest will be to manually grab the files from the known good commit, so my suggestion is the following:

* Start over at `feature/gemcalib`
```
git co -b gemcalib-cami-fixup aa78f4a
```
* `git cherry-pick` the commit that fixes the the `XHAL_ROOT` variable
```
git cherry-pick bcb719e
```
* Check out the relevant files from the unmerged commit of this PR (I got this list from inspecting the diff with `git diff --name-status ..fac3430` and selecting the files that should be included)
```
git co fac3430 Makefile
git co fac3430 gembase/html/css/bootstrap.css
git co fac3430 gembase/html/css/bootstrap.min.css
git co fac3430 gembase/html/scripts/bootstrap.js
git co fac3430 gembase/html/scripts/bootstrap.min.js
git co fac3430 gemcalibration/Makefile
git co fac3430 gemcalibration/html/scripts/cal_routines.js
git co fac3430 gemcalibration/include/gem/calib/Calibration.h
git co fac3430 gemcalibration/include/gem/calib/CalibrationWeb.h
git co fac3430 gemcalibration/src/common/Calibration.cc
git co fac3430 gemcalibration/src/common/CalibrationWeb.cc
git co fac3430 gemcalibration/xml/gemcalibration.xml
```
* Then patch `gemcalibration/Makefile` with the following (fixes the `Python.h` include problem you had previously):
```diff
diff --git a/gemcalibration/Makefile b/gemcalibration/Makefile
index b97d715..97d5b3b 100644
--- a/gemcalibration/Makefile
+++ b/gemcalibration/Makefile
@@ -7,8 +7,8 @@ ShortProject=gem
 Package=gemcalibration
 LongPackage=gemcalibration
 
-ShortPackage=gemcal
-PackageName=gemcalibration
+ShortPackage=calib
+PackageName=calib
 
 GEMCALIBRATION_VER_MAJOR=0
 GEMCALIBRATION_VER_MINOR=2
@@ -25,9 +25,12 @@ DynamicLibrary=gemcalibration
 IncludeDirs+=$(BUILD_HOME)/$(Project)/$(Package)/include
 IncludeDirs+=$(BUILD_HOME)/$(Project)/gembase/include
 IncludeDirs+=$(BUILD_HOME)/$(Project)/gemutils/include
-IncludeDirs+=/opt/xhal/include
+IncludeDirs+=$(XHAL_ROOT)/include
 
-DependentLibraryDirs+=/opt/xhal/lib
+UserCFlags  +=$(PYTHONCFLAGS)
+UserCCFlags +=$(PYTHONCFLAGS)
+
+DependentLibraryDirs+=$(XHAL_ROOT)/lib
 DependentLibraries+=xhal
 DependentLibraries+=boost_iostreams
```

From here you can either push this new branch and make a superseding PR, or you can force push that branch to the branch corresponding to this PR.
In either case, make sure that what that new branch is, corresponds to what you expect

jsturdy commented:
Superseded by #275
